storage: document ProgressNotify from storage.ListOptions

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
At first glance, it seems that the fields `storage.ListOptions.ProgressNotify` and `storage.ListOptions.Predicate.AllowWatchBookmarks`
are the same. Unfortunately, this is not the case.

This PR documents the differences and motivations for why these fields are actually distinct.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

p0lyn0mial commented:
/assign @wojtek-t

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/lister_watcher.go:
```diff
@@ -67,11 +67,12 @@ func (lw *listerWatcher) List(options metav1.ListOptions) (runtime.Object, error
 // Implements cache.ListerWatcher interface.
 func (lw *listerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	pred := storage.Everything
+	pred.AllowWatchBookmarks = true
```
I don't understand this change - you're not passing pred anywhere...

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go:
```diff
@@ -296,7 +296,7 @@ func (wc *watchChan) startWatching(watchClosedCh chan struct{}) {
 		opts = append(opts, clientv3.WithPrefix())
 	}
-	if wc.progressNotify {
+	if wc.internalPred.AllowWatchBookmarks {
```
Maybe I'm a bit paranoid, but I'm slightly worried about this change.

Up until now we have AllowWatchBookmarks and ProgressNotify separated.
This meant that:
- user could request bookmarks
- but if watchcache was disabled and that was going directly to etcd3 layer, it was ignored (and progress notify events were not returned back to the user).

With this change, the above behavior changes.
With watchcache, we send bookmarks ~every 1m, but progress notify are configured to lower values at least in some environments (like 5s).
This means, that if someone has watchcache disabled, they will suddenly may start getting 10x+ bookmarks. (which is obviously overhead on control-plane too).

At the very least, this will be the change for Events, for which watchcache is by default disabled.

I would like to see a convincing rationale why this is ok...

p0lyn0mial replied:
> but if watchcache was disabled and that was going directly to etcd3 layer, it was ignored (and progress notify events were not returned back to the user).

yes, I thought this is a bug :) I didn't know it was intentional.

Today when you do `k get --raw '/api/v1/namespaces/default/events?watch=true&allowWatchBookmarks=true'` against a cluster built from the `master` branch you won't get any bookmark events.

wojtek-t replied:
It's not necessary a bug. From this KEP:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/956-watch-bookmark#proposal

"
We consciously make it just a boolean flag - this gives kube-apiserver an ability to choose when they should be send without setting any expectations on user side how frequently and when it would be happening. In particular client isn't guaranteed to get any bookmarks. 
"

p0lyn0mial replied:
ok, do you see value in adding a mechanism similar to the one we have in the cache for bookmarks? 

I think it would be nice to have feature parity between different implementations.

p0lyn0mial replied:
yes, we already have different semantics depending on the caller:
- if watchcache then same-as-above
- if an api client then don't send any bookmark events

in addition to that we are about to add one more variation to the above list:
- if an api client requesting streaming then send only one special bookmark event


Let me bring back the progressNotify field add add some documentation if that is okay with you.

p0lyn0mial replied:
pushed, PTAL.

p0lyn0mial replied:
I have also opened https://github.com/kubernetes/kubernetes/issues/120348

jiahuif commented:
/triage accepted

p0lyn0mial committed:
storage: document ProgressNotify from storage.ListOptions

At first glance, it seems that the fields storage.ListOptions.ProgressNotify and storage.ListOptions.Predicate.AllowWatchBookmarks
are the same. Unfortunately, this is not the case.

This PR documents the differences and motivations for why these fields are actually distinct.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go:
```diff
@@ -71,13 +71,15 @@ func (f AttrFunc) WithFieldMutation(fieldMutator FieldMutationFunc) AttrFunc {
+	Limit       int64
+	Continue    string
+	// AllowWatchBookmarks when set indicates that the client
+	// want to receive bookmark events.
```
I would say that this particular field is self-explaining, so I would remove this comment.

p0lyn0mial replied:
Once you understand the meaning of that field, of course it is :)

I'm actually considering making it even more explicit, like this:
```
// AllowWatchBookmarks, when set, indicates that the end-client
// wants to receive optional bookmark events.
//
// As of today, this will be set when the 'allowWatchBookmarks'
// query parameter is set.
```

In the end, having more documentation won't harm, right? Thoughts?

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go:
```diff
@@ -71,13 +71,18 @@ func (f AttrFunc) WithFieldMutation(fieldMutator FieldMutationFunc) AttrFunc {
+	// wants to receive optional bookmark events.
+	//
+	// As of today, this will be set when the 'allowWatchBookmarks'
+	// query parameter is set.
```
I'm still against this comment:
- if you know what "watch bookmarks" are, then the name is self-explonatory, if you don't know what watch bookmarks are, this comment doesn't help anyway
- this is fairly low in the stack so talking about "end-client" doesn't make sense, it could be set by other layers of the stack and caught or sth...
- at this level, we shouldn't be talking abotu "query parameters" as these aren't really relevant here

So I still don't think this comment provides value...

p0lyn0mial replied:
ok, removed, PTAL.

p0lyn0mial committed:
storage/etcd: add TestWatchDispatchBookmarkEvents unit test

wojtek-t commented:
/lgtm
/approve
